run: Flesh out a bare implementation of `jj run`.

PhilipMetzger opened the pull request:
This is basically a MVP based on `fix`, caching is not implemented yet. The core functionality is in `run_inner()` and `rewrite_commit()`.

This is a draft to get further input from @arxanas and @hooper and all interested (builds since the 17.9.2024) . Also progress on #1869.  

TODO: rewrite trees and expose all files

<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the branch rather than adding commits on top. Use force-push when
pushing the updated branch (`jj git push` does that automatically when you
rewrite a branch). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

PhilipMetzger committed:
run: Add a minimal implementation of `jj run`

This is basically a MVP based on `fix`, caching is not implemented yet.
The core functionality is in `run_inner()` and `rewrite_commit()`.

It also exposes some initial environment variables for scripting purposes.

Stdout and stderr from the commands are buffered then output by the
parent process, matching the behavior of `jj fix`.

jennings committed:
run: Reveal command in CLI documentation

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+    Ok(results)
+}
+
+/// The result of a single command invocation in `run_inner`.
+struct RunJobResult {
+    /// The old `CommitId` of the commit.
+    old_id: CommitId,
+    /// The new `CommitId` for the commit.
+    rewritten_id: CommitId,
+    /// The new tree generated from the commit.
+    new_tree: Tree,
+}
```
Does this make sense?

arxanas replied:
question: Under what circumstance is `new_tree` different from the tree for the `rewritten_id` commit?

PhilipMetzger replied:
It never will be different, but to make my question a bit preciser should the caller or the callee do the actual "rewrite" of the commits tree?

arxanas replied:
I think it makes sense for the callee/result of the individual job to construct the new tree for an individual commit (considering that working-copy snapshotting happens as part of most commands anyways, it's probably not surprising or too expensive). And then the caller can choose to rewrite the commit structure with the new trees, if so desired.

For similar organizational purposes, I think the callee should be responsible for procuring the working copy to operate on.

PhilipMetzger replied:
Ok, I've removed the `rewritten_id` field, so the caller is responsible for creating it.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+    commits: &[StoredCommit],
+) -> Result<(), RunError> {
+    let mut command_futures = JoinSet::new();
+    for commit in commits {
+        command_futures.spawn(rewrite_commit(tx, commit.deref(), shell_command));
+    }
```
this should only run `jobs` amount of processes at once, not all.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -53,17 +303,84 @@ pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(
     // Fallback to a single user-visible job.
     .unwrap_or(1usize);
+
+    let mut done_commits = HashSet::new();
```
Does this make sense to track the done commits for the UI?

arxanas replied:
It seems fine; as opposed to what?

PhilipMetzger replied:
any other variant, since we want to integrate this later with [superconsole](../../../../facebookincubator/superconsole)?

PhilipMetzger replied:
> One of the things I would like to see here is an extension point for custom output/status aggregation.

Ok, I'll add a TODO. I assume this is meant to stream the output potentially across the network, if I understand you correctly?

PhilipMetzger replied:
I've added the TODO.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+                .map(|(i, s)| {
+                    // Incredibly "smart" way to say, append "-" if the number is odd "+"
+                    // otherwise.
+                    if i & 1 != 0 {
```
```suggestion
                    if i % 2 != 0 {
```

PhilipMetzger replied:
I'll leave it as that since that was its explained above.

arxanas replied:
Why not write it the conventional way so that we can get rid of the comment?

PhilipMetzger replied:
done

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+                    }
+                    s.to_owned()
+                })
+                .collect_vec();
```
suggestion: IIRC you can `collect` directly into a `String` without having to first create an intermediate `Vec`.

PhilipMetzger replied:
done

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+                        s.push('+');
+                    }
+                    s.to_owned()
+                })
```
issue: I think this leaves a `+` or `-` at the end of the full string, which is probably not intended? I assumed the `+`/`-`s should only be between tree IDs.

PhilipMetzger replied:
> issue: I think this leaves a `+` or `-` at the end of the full string, which is probably not intended?

This is true after some local testing and should be something to avoid,

PhilipMetzger replied:
This path is now only hit if we have multiple sides, so for simple commits we should only get hex ids. I'm not sure if that's correct though.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+                })
+                .collect_vec();
+            let mut obfuscated: String = ids.concat();
+            // `PATH_MAX` could be a problem for different operating systems, so truncate
+            // it.
+            if obfuscated.len() >= 255 {
+                obfuscated.truncate(200);
+            }
+            obfuscated
```
issue: I would rather just have the filesystem operation fail, rather than silently truncate the full path. In particular, it would probably be possible to add more sides to a conflict and then have two unrelated trees get assigned the same path on disk.

PhilipMetzger replied:
This works for *Nix based systems but for Windows where [this option](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry#enable-long-paths-in-windows-10-version-1607-and-later) is not set, so it'll make the command unusable. So maybe there's a better solution?

PhilipMetzger replied:
I commented out the code for now. If that is something we want, I'll remove it.

PhilipMetzger replied:
its deleted now.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+
+fn get_runtime(jobs: usize) -> tokio::runtime::Handle {
+    let mut builder = Builder::new_multi_thread();
+    if cfg!(watchman) {
+        // Watchman requires a multithreaded runtime, so just reuse it.
+        return runtime::Handle::current();
+    }
```
issue: I would delete this, and not bother trying to coordinate the runtime with the Watchman subsystem. We don't want to have the code here silently depend on invariants of the code elsewhere. If it's a real problem, we'll have the caller pass in the runtime.
issue: I don't think this instantiates the runtime with the right number of jobs anyways.

PhilipMetzger replied:
removed.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+        // Watchman requires a multithreaded runtime, so just reuse it.
+        return runtime::Handle::current();
+    }
+    if jobs == 1 {
+        builder.max_blocking_threads(1);
+    } else {
+        builder.max_blocking_threads(jobs);
+    }
```
```suggestion
    builder.max_blocking_threads(jobs);
```

PhilipMetzger replied:
done

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+// TODO: Create a caching backend, which creates these on a dedicated thread or
+// threadpool.
+struct StoredCommit {
+    /// Obfuscated name for an easier lookup. If a tree/directory its not set
```
question: Easier lookup by whom, for what purpose?

PhilipMetzger replied:
We need to replace the commits stored under `.jj/run/default/` if we have commits > jobs and that in-place. So the canonical commit is the initially created one, while those who overwrite it won't have a name. If that makes sense.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+    output_dir: PathBuf,
+    /// Self-explanatory
+    working_copy_dir: PathBuf,
+    state_dir: PathBuf,
```
question: What's the difference between `state_dir` and `output_dir`?

PhilipMetzger replied:
output is for other artifacts which aren't associated with the commit, e.g `stdout/stderr` and other files if necessary.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+    Ok((stdout, stderr))
+}
+
+fn create_working_copies(
```
suggestion: I would not preemptively create the working-copies; instead, I would invoke a function to create a working-copy on demand as part of running the job itself.

- We may as well benefit from the parallelization that `run` itself would use, rather than having to write it in two places.
- Extensions to `run` might make it so that we don't actually use all the working copies we allocated up-front.

PhilipMetzger replied:
Ack, but let's first get it running somewhat.

PhilipMetzger replied:
I'd still defer this unitl some work from @hooper has happened.

hooper replied:
A lot of this will shake out when we devise the abstractions for varying parallelism, execution order, caching, etc. This PR is probably too early to put much effort into that.

I think working copy creation will ultimately be triggered by the part that determines parallelism and execution order, so that we can hide the latency sometimes while also deferring the work until it is almost needed. Having it in the inner loop may make sense until then, although I wonder if the parallel access pattern would make it slower.

PhilipMetzger replied:
> I think working copy creation will ultimately be triggered by the part that determines parallelism and execution order, so that we can hide the latency sometimes while also deferring the work until it is almost needed.

+1 from me for the general plan. 

> Having it in the inner loop may make sense until then, although I wonder if the parallel access pattern would make it slower.

I think a fan-out option makes sense if everything is accessed through the network, this also wouldn't hurt the planned on-disk implementation as much.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+}
+
+/// The result of a single command invocation in `run_inner`.
+struct RunJobResult {
```
suggestion (non-blocking): Try to name this without using `Result`, since it it's not a `Result`.

PhilipMetzger replied:
renamed to `RunJob`.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+    Ok(results)
+}
+
+/// The result of a single command invocation in `run_inner`.
```
suggestion: Don't refer to `run_inner` in this comment as it's primarily an implementation detail. You can say something like that it's the result of running the provided command on a single commit.

PhilipMetzger replied:
done

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+}
+
+/// Rewrite a single `StoredCommit`.
+async fn rewrite_commit<'a>(
```
issue: This doesn't actually rewrite any commits, so it shouldn't be called `rewrite_commit`.

PhilipMetzger replied:
it does rewrite the commits tree and I want to rewrite it in there if possible.

PhilipMetzger replied:
imo, it does the core part now, so the name could stay.

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -14,12 +14,262 @@
+/// A commit stored under `.jj/run/`
+// TODO: Create a caching backend, which creates these on a dedicated thread or
+// threadpool.
+struct StoredCommit {
```
question: In what sense is this a "stored" commit? All commits are stored. The name/comment don't explain the intention of this struct, just a detail about that it's persistent in some fashion.

PhilipMetzger replied:
Renamed to `OnDiskCommit`, maybe that is better?

arxanas commented on cli/src/commands/run.rs:
```diff
@@ -53,17 +303,84 @@ pub fn cmd_run(ui: &mut Ui, command: &CommandHelper, args: &RunArgs) -> Result<(
     // 2. the amount of cores available.
     // 3. a single job, if all of the above fails.
-    let _jobs = match args.jobs {
+    let jobs = match args.jobs {
```
issue: Isn't the convention that `0` means "use all available parallelism"? I think we should do that instead of producing an error here.

PhilipMetzger replied:
See #4149

PhilipMetzger commented:
This builds now, but fails when invoking the command with a "failed to write, no such fie or directory" somewhere.

PhilipMetzger reviewed (commented):
I've added some replies since @hooper brought this just up in the Discord.

PhilipMetzger committed:
run: Add `revsets.run` to give `jj run` a default revset

So each command no longer requires the `--revisions` by default. This is
similar to `jj fix`'s `revsets.fix`, so maybe we should unify them at
some point.

AFAIK, hooper and arxanas have expressed interest in that.

PhilipMetzger commented:
This is the state from the [Discord message][dc], PTAL. Its just missing tests to land in the February release. 


[dc]: https://discord.com/channels/968932220549103686/969291218347524238/1309224061464743978

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,15 +14,375 @@
+        id = stored_commit.commit.id().hex(),
+        shell_command = shell_command.as_str()
+    );
+    let mut command = tokio::process::Command::new(get_shell_executable())
+        // TODO: This is not portable as Windows expects a "/c" instead of a "-c".
+        .arg("-c")
```
I consider this a problem.

PhilipMetzger replied:
it is now done in a cross-platform way.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,15 +14,375 @@
+        );
+    }
+
+    while let Some(res) = command_futures.join_next().await {
+        let done = res.unwrap().expect("should not fail joining a job");
```
I ran into a panic here when testing locally with `jj run 'sed/some/string/g <file>'`, as the last file was unaffected which returned a exitcode of 255 (2).

PhilipMetzger replied:
also adjusted, but the tests still fail.

PhilipMetzger reviewed (commented):
Some minor discussion points.

martinvonz commented on cli/tests/test_run_command.rs:
```diff
@@ -0,0 +1,87 @@
+
+use crate::common::TestEnvironment;
+
+/// Sets up a repo with three initial commits, containing some text.
```
nit: description seems stale (it only creates a single commit - for the working copy)

PhilipMetzger replied:
it's adjusted now.

martinvonz commented on cli/tests/test_run_command.rs:
```diff
@@ -0,0 +1,87 @@
+            "'mutable()..@'",
+        ],
+    );
+    // all commits should be modified
```
this doesn't seem to test that

ilyagr reviewed (commented):
I haven't thought about the tests or the code yet, but I tried using it and - lo and behold - it works! It rebases cleanly on top of main (except for the changelog)! I can now rename a file in a stack! 🎉 

I did make a bunch of very minor corrections that might be easiest to show as a patch, they are pretty self-evident (except for the first one, that simply reverts the change to the CHANGELOG). Some of them are clippy-related, which might be a bit early, but I thought I'd share them anyway. Others fix some simple thing I discovered while trying to use this version of `jj run`.

https://github.com/jj-vcs/jj/commit/ig-pr4021

You can pull the branch or use

https://github.com/jj-vcs/jj/commit/ig-pr4021.diff

ilyagr commented on cli/src/commands/run.rs:
```diff
@@ -14,15 +14,376 @@
+    let mut command = tokio::process::Command::new(prog)
+        .arg(first_arg)
+        .arg(shell_command.as_str())
+        // set cwd to the working copy directory.
```
(Possibly for later)

For `jj run` command, it is surprising to me that commands are run in the working copy directory instead of the subdirectory I am in. For example, I think it's natural to be in the `cli/src` directory and to run `jj run "mv config.rs somethingelse.rs" -r ...`.

I think this should be governed by an option to `jj run`, and the default should be to run the command in the subdirectory. Until this is implemented, you could make `jj run` vocal about the fact that it's running the commands in the working copy dir.

PhilipMetzger replied:
I don't think its for discussion here, as it out of scope for this MVP. Please add it either to #1869 or #3808

PhilipMetzger replied:
This TODO is related to this request from @matts1: 
> No mention was made about the working directory, so I'm curious how relative paths would work
IMO, we should set the $JJ_REPO environment variable so that a relative path to the root of the repo can be $JJ_REPO/foo
Maintaining relative paths has value
A unix mount namespace could solve this by literally just overlaying our new virtual file system on top of the real one
But unclear how this'd work on non-unix systems

jennings replied:
I added this. I think it's likely someone will try `jj run 'rustfmt file_in_this_dir.rs'` and expect it to work (nevermind that that particular thing should probably be done with `jj fix`).

PhilipMetzger commented:
> I did make a bunch of very minor corrections that might be easiest to show as a patch, they are pretty self-evident (except for the first one, that simply reverts the change to the CHANGELOG). Some of them are clippy-related, which might be a bit early, but I thought I'd share them anyway. Others fix some simple thing I discovered while trying to use this version of `jj run`.

I already fixed those clippy lints locally, but thanks :-). 

> The same happens with `mkdir -p qq`, though I think either should work (if there is no `qq`).
> 
> **Update:** It might have little to do with the command, it might just have to do with something else. Maybe revisions with different commit ids but the same change ids as revisions the commmand was run before? I should probably take a look at the tests :)
> 
> In any case, ` cargo run -- run 'true' -r '@|@-'` results in the same error.
> 
> **Update 2:** A possible reason is that commit `@` was empty. Running `rm -r .jj/run/` seems to be a workaround.

^, is the error I face with the tests.

ilyagr commented:
> I already fixed those clippy lints locally, but thanks :-).

Oh, you can update the PR then. Did you also get these parts? They jumped at me when I tried to use `jj run`. (I also found one more I just added)

```diff
diff --git a/cli/src/commands/run.rs b/cli/src/commands/run.rs
index 9ccb313eba..c184cc2a4e 100644
--- a/cli/src/commands/run.rs
+++ b/cli/src/commands/run.rs
@@ -361,7 +361,7 @@
     let (dirty, _) = tree_state.snapshot(&options).unwrap();
     if !dirty {
         tracing::debug!(
-            "commit {} was not modified as the passed command did nothing",
+            "commit {} was not modified as the passed command did not modify any tracked files",
             commit.id()
         );
     }
@@ -404,7 +404,7 @@
     shell_command: String,
     /// The revisions to change.
     #[arg(long, short, default_value = "@", value_name = "REVSETS")]
-    revisions: RevisionArg,
+    revisions: Vec<RevisionArg>,
     /// A no-op option to match the interface of `git rebase -x`.
     #[arg(short = 'x', hide = true)]
     exec: bool,
@@ -417,7 +417,7 @@
     let mut workspace_command = command.workspace_helper(ui)?;
     // The commits are already returned in reverse topological order.
     let resolved_commits: Vec<_> = workspace_command
-        .parse_revset(ui, &args.revisions)?
+        .parse_union_revsets(ui, &args.revisions)?
         .evaluate_to_commits()?
         .try_collect()?;
     // Jobs are resolved in this order:
@@ -482,7 +482,7 @@
         // Yeet everything, caching is better implemented in a follow-up.
         fs::remove_dir_all(&run_path)?;
 
-        println!("No commits were rewritten as the command did nothing");
+        println!("No commits were rewritten as the command did not modify any tracked files");
         tx.finish(
             ui,
             format!("run: No-op on {visited} commits with {shell_command}"),
```

PhilipMetzger commented on CHANGELOG.md:
```diff
@@ -16,6 +16,10 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 ### New features
 
+* A minimal version of `jj run` was implemented.
+  Follow [#1869](https://github.com/jj-vcs/issues/1869) for further
```
done

nightscape commented:
I have been needing this quite frequently. @PhilipMetzger thank you for your efforts!
Could you give a high-level overview over what is missing for an initial version?

jennings commented:
With Philip's blessing, I amended and pushed a new revision with these changes:

- Feature changes
  - Default revset is now `reachable(@, mutable())`, which aligns with `jj fix`. I think `@` was a surprising default because if that's what you wanted, you'd just run the command directly without `jj run`.
  - `run` subcommand is now shown in `jj -h` output
  - stdout and stderr from subprocesses is buffered and emitted all at once by the parent process. This allows scripts to output information and see them in the terminal or pipe them somewhere.
  - Execute commands from the same subdirectory of the working copy where `jj run` is run from.

- Implementation changes
  - Run workspaces are now keyed by commit ID instead of tree ID. An empty commit shares a tree ID with its parent, so they were colliding.
  - Close stdin on spawned subprocesses
  - `writeln!(ui.stderr(), ...)` instead of `println!(...)`
  - Tests are now passing, including the one that uses `cargo fmt`.
  - Reworded the commit message to say "minimal implementation", matching the CHANGELOG entry.

I haven't yet gone through open comments in this PR, I'll do that soon, hopefully.

jennings commented:
I _think_ all the comments people brought up have been answered or addressed. Kind of hard to tell because GitHub loses comment threads after a force push :(

jennings commented:
I think the first improvement I want to make after this is to have a pool of workspaces that are reused for multiple commits, and not deleted between invocations of `jj run`.

That doesn’t seem too hard to do if I only consider the one `jj run` instance, but if someone runs concurrently from two different terminals, there needs to be some kind of cross-process lock.

PhilipMetzger commented:
> Feature changes
> 
> * Default revset is now `reachable(@, mutable())`, which aligns with `jj fix`. I think `@` was a surprising default because if that's what you wanted, you'd just run the command directly without `jj run`.

If you add that please just copy the settings reading from `fix` in a separate commit. 

> * Reworded the commit message to say "minimal implementation", matching the CHANGELOG entry.

can you please leave my stuff as is when you're saying you're fixing them. I don't see much value in this. 

You can also just squash the commit which makes it visible.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
-use crate::command_error::user_error;
+use crate::command_error::CommandErrorKind;
 use crate::ui::Ui;
 
-/// (**Stub**, does not work yet) Run a command across a set of revisions.
+#[derive(Debug, thiserror::Error)]
+enum RunError {
+    #[error("failed to checkout the commit {}", .0)]
+    FailedCheckout(CommitId),
+    #[error("the command '{}' failed with {} for commit {}", .0,.1, .2)]
+    CommandFailure(String, ExitStatus, CommitId),
+    #[error(transparent)]
+    IoError(#[from] io::Error),
+    #[error("failed to create path {} with {}", .0.to_string_lossy(), .1)]
+    PathCreationFailure(PathBuf, io::Error),
+    #[error("failed to load a commits tree")]
+    TreeState(#[from] TreeStateError),
+    #[error(transparent)]
+    Backend(#[from] BackendError),
+    #[error(transparent)]
+    JobFailure(#[from] JoinError),
+}
```
nit: missing documentation

PhilipMetzger commented on cli/src/commands/run.rs:
````diff
@@ -14,71 +14,588 @@
-/// This allows pre-commit integration and other funny stuff.
+/// $ jj run 'pre-commit run .github/pre-commit.yaml' -j 4
+/// ```
+// TODO: align interface with `jj bisect run --`
````
nit: address this because my local patch already does this.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
         long = "revision",
         short,
-        default_value = "@",
+        default_value = "reachable(@, mutable())",
```
nit: drop and move to a separate commit copying the settings reading from `fix`.

jennings replied:
I think we may as well come out of the gate with a useful revset as a default. @ isn't useful because you don't need `jj run` for that.

PhilipMetzger replied:
> I think we may as well come out of the gate with a useful revset as a default. @ isn't useful because you don't need `jj run` for that.

I will just add that commit here, since you want that. 

Edit: And requiring that people need to use `-r <revset>` to start out is not bad.

PhilipMetzger replied:
I've pushed the commit and deleted the respective change from you there.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
     /// How many processes should run in parallel, uses by default all cores.
     #[arg(long, short)]
     jobs: Option<usize>,
+
+    /// Run the command from the working-copy root in each commit instead of
+    /// from the subdirectory `jj run` was invoked from.
+    #[arg(long)]
+    root: bool,
```
nit: separate PR please

jennings replied:
Done.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
+
+    workspace_command
+        .check_rewritable(resolved_commits.iter().ids())
+        .await?;
```
nit: this should just be the `chec_rewritable` check. Matching the interface of every other command

jennings replied:
Yeah you're probably right. I kind of thought about it later but forgot to come back to it.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
+    let rt = get_runtime(jobs);
+    // TODO: Add a extension point for custom output/status aggregation.
+    let mut done_commits = HashSet::new();
+    let (sender_tx, mut receiver) = mpsc::channel(resolved_commits.len());
+
+    // Run each command from the subdirectory the user invoked `jj run` from,
+    // unless `--root` overrides that. The subdir is relative to the workspace
+    // root, which is canonical (per `CommandHelper::cwd` docs).
+    let subdir = if args.root {
+        PathBuf::new()
+    } else {
+        command
+            .cwd()
+            .strip_prefix(workspace_command.workspace_root())
+            .map(Path::to_path_buf)
+            .unwrap_or_default()
+    };
```
nit: ditto with separate PR

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
+        }
+        if visited == stored_len {
+            break;
+        }
```
nit: I don't think you've understood me here, they still should write to their respective stderr/stdin files.

jennings replied:
I did, but I think it makes more sense to emit stdout and stderr to the parent process's stdout/stderr by default. My reasoning:

- It's what `git rebase` does, so people may expect it. I know "git does it" isn't a good reason to do anything, but "people expect it" is.
- I think it's more useful than writing to files buried in `.jj/run/default/*` that get deleted on a successful run.
- If the user wants to write them to files instead, they can add `> path/to/stdout 2> path/to/stderr` in their command.

PhilipMetzger replied:
> I did, but I think it makes more sense to emit stdout and stderr to the parent process's stdout/stderr by default. My reasoning:
> 
> * It's what `git rebase` does, so people _may_ expect it. I know "git does it" isn't a good reason to do anything, but "people expect it" is.

Sorry I don't think "people expect it" in relation to a new command which loosely relates to a `git rebase -x` makes sense here. I'm quite sure that if people want that opt-in we should provide it, just not now. 

> * I think it's more useful than writing to files buried in `.jj/run/default/*` that get deleted on a successful run.

You know this was temporary for https://github.com/jj-vcs/jj/pull/4021#issuecomment-4571824408 and to cut down the size of this. Its a better experience if we can provide the failing `stdout/stderr` if the command fails on an individual commit. 

> * If the user wants to write them to files instead, they can add `> path/to/stdout 2> path/to/stderr` in their command.

I don't want to support this currently. If you open FR for it we can support that via a flag but its also out of scope for this PR.

martinvonz replied:
> I think it makes more sense to emit stdout and stderr to the parent process's stdout/stderr by default.

Since `jj run` will allow the command to be run on multiple commits concurrently, that would result in interleaved output from many commands, right? I think we want to capture the output and present it to the user in a more structured way, though I don't know what that would be. @hooper: what does `hg run` do? Hmm, looking at the (internal) code, it actually looks like it interleaves the output. Is that correct? I know you made the `hg blaze` command print output in a nice way. Have you thought about how a generic `jj run` should present the output?

jennings replied:
> Since jj run will allow the command to be run on multiple commits concurrently, that would result in interleaved output from many commands, right?

This PR does ensure they're not interleaved, I did essentially what `jj fix` does with stderr: Buffer the stdout/stderr from each process, then as they complete they have to acquire a lock on `ui.stdout()`/`ui.stderr()` to emit it all at once.

My thought is that if the user wants structured data, they'll output one line per commit, then pipe it to something that can parse that. For example, to figure out which commits have passing tests:

```command
$ jj run -- bash -c 'cargo test >/dev/null;echo "$JJ_COMMIT_ID,$?"'
aaaaa,0
bbbbb,1
ccccc,0
```

Or, their command can write structured data to files elsewhere in the system, then they'll know where to read it from after `jj run` exits.

jennings replied:
Since the current behavior matches hooper’s recommendation, I’m going to resolve this thread. We can revisit it in a new PR if necessary.

PhilipMetzger commented on cli/tests/test_run_command.rs:
```diff
@@ -0,0 +1,332 @@
+    [EOF]
+    ");
+
+    // The rewritten `with-sub` commit has `sub/ran.txt`, alongside the
+    // pre-existing `sub/file.txt`.
+    insta::assert_snapshot!(
+        work_dir.run_jj(&["file", "list", "-r", "@-"]),
+        @r"
+    seed.txt
+    sub/file.txt
+    sub/ran.txt
+    [EOF]
+    "
+    );
+}
+
+#[test]
+fn test_run_root_flag() {
+    let test_env = TestEnvironment::default();
+    test_env.run_jj_in(".", ["git", "init", "repo"]).success();
+    let work_dir = test_env.work_dir("repo");
+    work_dir.write_file("sub/file.txt", "x");
+    work_dir.run_jj(&["commit", "-m", "with-sub"]).success();
+
+    // Invoke `jj run` from inside sub/, but pass `--root` so the command
+    // executes from the workspace root and `ran.txt` lands at the top level.
+    let sub_dir = work_dir.dir("sub");
+    sub_dir
+        .run_jj(&["run", "--root", "touch ran.txt", "-r", "@-"])
+        .success();
+
+    insta::assert_snapshot!(
+        work_dir.run_jj(&["file", "list", "-r", "@-"]),
+        @r"
+    ran.txt
+    sub/file.txt
+    [EOF]
+    "
+    );
+}
```
nit: separate PR

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
+        // keyed by commit id to keep paths unique; multiple commits in the input
+        // may share a tree (e.g. the working-copy commit immediately after `jj
+        // commit`), so a tree-keyed name would collide.
+        let commit_path = base_path.join(commit.id().hex());
```
I personally don't like this change at all.

jennings replied:
For what reason? I couldn't find the conversation about it to confirm, but I think this may have been the root cause for the `stdout`/`stderr` file contention that tests were running into.

PhilipMetzger replied:
> For what reason?

It allows users to introspect the commits as soon as they stay permanently on-disk which is not something I want to support at all. 

> I couldn't find the conversation about it to confirm, but I think this may have been the root cause for the `stdout`/`stderr` file contention that tests were running into.

Again the issue I ran into is that two separate `run` processes started by `nextest` wanted to create a file which potentially already existed on disk. The naming scheme had no influence on this.

PhilipMetzger commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,588 @@
+    stdout: Vec<u8>,
+    /// Bytes the subprocess wrote to its stderr, captured in full.
+    stderr: Vec<u8>,
+    /// True if the command wasn't run because the per-commit working directory
+    /// (the subdirectory `jj run` was invoked from) didn't exist in this
+    /// commit's tree.
+    skipped: bool,
```
nit: also something for later

jennings replied:
It's already done and I think we want this behavior, don't we? EDIT: Oh are you just asking to split the working directory behavior to a new commit?

PhilipMetzger replied:
> It's already done and I think we want this behavior, don't we? EDIT: Oh are you just asking to split the working directory behavior to a new commit?

No I'm just saying that is also a separate PR, for some reason you constantly choose to blow the scope of this instead of trying the minimal thing.

jennings replied:
The `--root` flag and `skipped` concept have been extracted to a separate commit on top of the minimal implementation commit.

PhilipMetzger commented on cli/src/commands/mod.rs:
nit: squash this commmit

PhilipMetzger reviewed (commented):
As the original author I really don't like the changes.

jennings commented:
I ended up making @arxanas's suggestion to create working copies on-demand rather than all up front. This removed the need for the OnDiskCommit struct.

jennings committed:
run: Run commands from the same working copy subdirectory

Run each command from the same subdirectory of the working copy that `jj
run` is run from.

Adds a `--root` option to instead run from the root of the working copy
regardless of where `jj run` is executed from.

PhilipMetzger commented on cli/src/commands/run.rs:
> So each command no longer requires a the `--revisions` by default.

Drop the "a"

thoughtpolice commented on CHANGELOG.md:
```diff
@@ -21,6 +16,10 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+* A minimal version of `jj run` was implemented.
+  Follow [#1869](https://github.com/jj-vcs/jj/issues/1869) for further
+  improvements.
```
The tense here is weird, nobody necessarily knows a priori what `run` is and I wouldn't bother mentioning a discussion issue tbh. I'd suggest:

> `jj run` allows you to run a command over a set of changes, each with their own private working copy; the commands may update the working copy and changes/conflicts are propagated accordingly e.g. `jj run 'cargo check --all-features'` or `jj run 'cargo fix'` behaves as one might expect.

Also, move this into the release highlights IMO.

jennings replied:
Fixed, thanks

thoughtpolice commented on cli/Cargo.toml:
```diff
@@ -103,6 +103,7 @@ shlex = { workspace = true }
 slab = { workspace = true }
 tempfile = { workspace = true }
 textwrap = { workspace = true }
+tokio = { workspace = true, features = ["fs", "process", "rt-multi-thread", "sync"] }
```
Put this in the main `Cargo.toml` workspace and then just inherit that feature set, please. We've let this slip a few times in `lib/` and should probably audit divergent feature flags like this, but AFAICS there's no need for it here.

jennings replied:
Fixed.

thoughtpolice commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,573 @@
+    for commit in commits.iter() {
+        command_futures.spawn_on(
+            rewrite_commit(
+                // TODO: handle/propagate error here
```
Old comment? This is handled by `join_next()`, is it not?

PhilipMetzger replied:
it should be moved to line 216

jennings replied:
I don't think it's clear what "handle/propagate error" means we want to eventually do, so if _anything_ I'd rather have a comment explaining the problem with the current approach. But I don't know what that problem is, we seem to behave sensibly right now: If a command returns nonzero, we abort the whole transaction. I'll  remove the comment unless we can specify what we'd like to do instead?

PhilipMetzger replied:
> I don't think it's clear what "handle/propagate error" means we want to eventually do, so if _anything_ I'd rather have a comment explaining the problem with the current approach.

This was initially written for cloning the `.gitignores` which may or may not exist when we created the snapshot options and the 64MB snapshot limit which should be read from the correct config. 

> But I don't know what that problem is, we seem to behave sensibly right now: If a command returns nonzero, we abort the whole transaction. I'll remove the comment unless we can specify what we'd like to do instead?

For a `-k --keep-going` flag we will need to ignore individual errors and not stop running processes but that's for later.

thoughtpolice commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,573 @@
+    args: Vec<String>,
+    /// Working directory for the subprocess, relative to the workspace root.
+    /// Empty means run from the workspace root.
+    subdir: PathBuf,
```
small nit: make this an `Option<PathBuf>` and then you don't even need the `is_dir()` check in the `--root` case below.

jennings replied:
Changed, thanks.

thoughtpolice commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,573 @@
+    base_path: &Path,
+    commit: &Commit,
+) -> Result<(PathBuf, TreeState), RunError> {
+    // Per-commit working-copy directory, keyed by commit id so concurrent jobs
+    // don't collide and so the working copy is reproducible across runs.
+    let commit_path = base_path.join(commit.id().hex());
+    // A previous `jj run` that failed before cleanup may have left this
+    // directory behind. Clear it so the working_copy/state subdirs below
+    // start from a clean slate.
+    if commit_path.exists() {
+        tracing::debug!(
+            dir = ?commit_path,
+            commit = commit.id().hex(),
+            "removing leftover directory from a previous run"
+        );
+        fs::remove_dir_all(&commit_path)
+            .map_err(|e| RunError::PathCreationFailure(commit_path.clone(), e))?;
+    }
```
Is this not just a massive TOCTOU race? If you run `jj run -- slow_cmd` and `jj run -- fast_cmd` over the same set of commits, would the fast command not just obliterate the slow commands' working copies while they're running, causing spurious failures?

jennings replied:
It would. My follow up "pooled working copies" implementation that I'm working on adds a lockfile at `.jj/run/pool/N/lock` so multiple processes don't stomp on each other, we could probably use that here. What would you like to do?
- Implement something now before merging
- Implement it as a (hopefully) fast follow-up PR

PhilipMetzger replied:
> lockfile at `.jj/run/pool/N/lock`

can this be under `.jj/run/default/pool/` instead so its easier to pull out a trait for it?

jennings replied:
For the current implementation, I would put it in `.jj/run/default/<commit_id>/lock`

(Adding the following for context, not for discussion) My current implementation for pooled working copies creates directories `.jj/run/pool/{1-N}`  with the same layout as `.jj/run/default/<commit_id>` has. You choose which set of directories to use by passing either `--strategy isolated` or `--strategy pool`. So, my thought was that this "default" directory is where `--strategy isolated` does its work and the "pool" directory is where `--strategy pool` does its work.

PhilipMetzger replied:
> For the current implementation, I would put it in `.jj/run/default/<commit_id>/lock`

If Austin wants that, SGTM. 

> (Adding the following for context, not for discussion) My current implementation for pooled working copies creates directories `.jj/run/pool/{1-N}` with the same layout as `.jj/run/default/<commit_id>` has. You choose which set of directories to use by passing either `--strategy isolated` or `--strategy pool`. So, my thought was that this "default" directory is where `--strategy isolated` does its work and the "pool" directory is where `--strategy pool` does its work.

These again are internal implementation details so putting them under `.jj/run/default/{strategy}/{commit-id}` should be fine.

jennings replied:
What does "default" mean in this path? What other values besides "default" might there be?

PhilipMetzger replied:
> What does "default" mean in this path?

It means that this is the default backend for run, since we want to give downstreams a way to extend it. 

> What other values besides "default" might there be?

I want to have a RE one built on https://bazel.build/remote/rbe and another one which is similar to Bazel's hybrid one, which races executions between a RE backend and the local one.

jennings replied:
I think what I was tentatively calling "strategy" is the same concept as the backend you're describing.

- The "isolated" backend creates per-commit fresh working copies, using `.jj/run/default` as its workspace (or `.jj/run/isolated`).
- The "pool" backend reuses working copies between invocations, using `.jj/run/pool` as its workspace.
- A "remote" backend either wouldn't need any local directories at all, or it would use `.jj/run/remote` as its workspace (which would _not_ contain working copies, but instead probably bookkeeping info).
- A "race" backend wouldn't need its own workspace, but would use both `.jj/run/pool` and `.jj/run/remote`.

This makes the `.jj/run/default/{strategy}` path seem more nested than necessary to me.

PhilipMetzger replied:
> I think what I was tentatively calling "strategy" is the same concept as the backend you're describing.
> 
> * The "isolated" backend creates per-commit fresh working copies, using `.jj/run/default` as its workspace (or `.jj/run/isolated`).
> * The "pool" backend reuses working copies between invocations, using `.jj/run/pool` as its workspace.

I mean you're first two strategies or "backends" use the same mechanism, that's why they should stay under `default` where all checked out working copies land. 

> * A "remote" backend either wouldn't need any local directories at all, or it would use `.jj/run/remote` as its workspace (which would _not_ contain working copies, but instead probably bookkeeping info).

A remote just needs a directory and a file with some internals in there. 

> * A "race" backend wouldn't need its own workspace, but would use both `.jj/run/pool` and `.jj/run/remote`.

A "hybrid" backend may choose to materialize when racing so having them having its own directories is a better design, since each backend is responsible for the management of the working copies. 

> This makes the `.jj/run/default/{strategy}` path seem more nested than necessary to me.

See my first point.

thoughtpolice replied:
Worrying about the location of paths like this right now is not relevant. We can cross that bridge when/if any further work on abstracting or pulling out traits is done, not before, this is just splitting hairs over nothing and delaying this PR.

FWIW, I do think it's necessary to fix this. But I would be completely fine if this PR just had a single global lock around a run invocation, so that only one `jj run` can execute at a time per repository. To be clear, that single `run` invocation could be using concurrent threads executing many things in parallel over a large number of commits. But only one `jj run` _command_ can be executing at a time. This would be similar to how `cargo` locking works today and would be enough to stop a massive footgun from happening.

jennings replied:
I added a single global lock at `.jj/run/default/lock` that is acquired before beginning execution.

thoughtpolice commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,573 @@
+        .args(&spec.args)
+        .current_dir(&exec_dir)
+        // TODO: relativize
+        // .env("JJ_PATH", working_copy_dir)
```
What does this comment mean? FWIW we could maybe get rid of it because you can just run a command that gives you `$CWD` in a shell or something?

PhilipMetzger replied:
This was a TODO left over for this: 

> No mention was made about the working directory, so I'm curious how relative paths would work
IMO, we should set the $JJ_REPO environment variable so that a relative path to the root of the repo can be $JJ_REPO/foo
Maintaining relative paths has value
A unix mount namespace could solve this by literally just overlaying our new virtual file system on top of the real one
But unclear how this'd work on non-unix systems

> This is another great point, which never was brought up. I agree that the $JJ_REPO env var should be included from the beginning.

This from #1869 which originally was https://github.com/jj-vcs/jj/issues/3577#issuecomment-2113627939. This is a user request which we can punt to later.

thoughtpolice replied:
That makes sense. We should probably just include `JJ_PATH` as it is and uncomment this line, then. We can handle any edge cases with relative paths coming up later.

jennings replied:
I added it as `JJ_WORKSPACE_ROOT` to match the env var that `jj util exec` provides. I can change it if that's not a good idea.

thoughtpolice reviewed (commented):
I have more nits here but we should really just get this over the line and put it into users hands at this point, IMO.

My comment below about the working copy races is the most important one. Maybe I misread something and it's a non-issue, but otherwise we definitively need to fix that.

PhilipMetzger commented on cli/src/commands/run.rs:
nit: missing changelog entry update

PhilipMetzger commented on cli/src/commands/run.rs:
nit: also needs a changelog update and I missed the `config-schema.json` update here

jennings replied:
I added the config-schema update, the changelog has been updated.

PhilipMetzger reviewed (commented):
minor comments

hooper commented:
I don't want to block this PR, so I will just mention a few things that will be important to Google based on our experience with `hg run`:
 - The ability to choose between "run sequentially in the working copy" and "run in parallel in remote/external working copies". We use a VFS to quickly materialize huge monorepo working copies, and they wouldn't be mounted under `.jj/run/`.
    - We also went with literally `--strategy foo` as a UI for this, but I think our users mostly go with the defaults we set for the command and its aliases (imagine something like `hg bazel foo` -> `hg run --strategy=magic_vfs_thing -j10 -- bazel foo` with some special handling for the progress output).
    - We generate these on demand, re-use them when reasonable, and have a cache of them. I'm happy to see consideration for these things already!
    - I guess we'll need a backend for registering custom/extension implementations of this so the user can choose between them. `create_working_copy` doesn't look like it supports that yet.
 - Passing output through to the terminal is a very useful default, and redirecting it to a file would ideally be possible via normal shell scripting techniques, even if there's another mechanism. Buffering/interleaving requires some care, and having a few reasonable behaviors available via flags/configs should be valuable. I just wouldn't want to over-index the implementation or UI on any one use case.
    - I also think -j1 is a good default, partly because it defers the output buffering/interleaving problem until the user asks for it by increasing parallelism. It can also mitigate a tendency for swapping or other kinds of resource exhaustion to be a foot gun.
    - When run on a tty, supporting interactive commands is useful.
 - The transform_descendants aspect is great, and much better than trying to have the user put something like `jj squash` into the command, so thank you for that. However, it should probably be possible to turn it off so the working copies can be treated as throwaway sandboxes. Some use cases might benefit a lot from avoiding the cost of checking for changes at the end, especially if caching allows them to avoid the setup cost too.
    - The option to use `--restore-descendants` would be useful for the same reasons `jj fix` always works that way.

Please take all of that as advisory, I apologize that I don't have much time to help with this feature right now.

jennings commented:
> I guess we'll need a backend for registering custom/extension implementations of this so the user can choose between them. `create_working_copy` doesn't look like it supports that yet.

It's not for this PR, but [I have an implementation](https://github.com/jj-vcs/jj/compare/push-mqxwqtnwnpts...jennings:jj:run-pool?expand=1) that lets you choose between `--strategy isolated` (this PR's behavior) and `--strategy pool` (reusable working copies). I'm hoping to follow up with that after this lands. 

> I also think -j1 is a good default, partly because it defers the output buffering/interleaving problem until the user asks for it by increasing parallelism. It can also mitigate a tendency for swapping or other kinds of resource exhaustion to be a foot gun.

Sounds good to me, I'll make this change, then in a follow-up maybe add a config to override the default.

> However, it should probably be possible to turn [amending] off so the working copies can be treated as throwaway sandboxes.

I think that makes sense. I think we can add `--no-amend` or something like that in a follow-up PR.
   
> The option to use `--restore-descendants` would be useful for the same reasons `jj fix` always works that way.

Agree.

thoughtpolice commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,594 @@
+
+/// Creates the required directories for a StoredWorkingCopy.
+/// Returns a tuple of (`working_copy` and `state`).
+async fn create_working_copy_paths(path: &Path) -> Result<(PathBuf, PathBuf), RunError> {
```
i think we can just inline this function into `create_working_copy` tbqh

thoughtpolice commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,594 @@
+    Ok((working_copy, state))
+}
+
+fn get_runtime() -> tokio::runtime::Runtime {
```
inline this too, it has only 1 call site.

thoughtpolice commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,594 @@
+/// Compute and ensure the base directory under which each per-commit working
+/// copy lives, then acquire an exclusive lock on it. Returns the base path and
+/// the held lock (drop to release).
+fn ensure_base_path(repo_path: &Path) -> Result<(PathBuf, FileLock), RunError> {
```
yeah, let's just inline this too. i think doing so is good because keeping the FileLock closer to where it's actual lifetime is (cmd_run) makes it easier to understand imo. You can also use `drop(lock)` IMO to get rid of the unused variable suppression and make the lock scope a little more clear, IMO.

thoughtpolice reviewed (approved):
These are all mostly just readability nits, I leave them to your better judgement if you want to address them. Otherwise I'm happy with how this first version looks; tysm for helping get this over the line Stephen, and Philip for starting it.

thoughtpolice commented:
Oh, and one more thing: I'd probably just merge the first two commits together (the one that adds the `--root` flag into the OG commit.) But that's just an idle thought, don't put too much stock in it.

mirkomartn commented on cli/src/commands/run.rs:
```diff
@@ -14,71 +14,512 @@
+    while let Some(res) = command_futures.join_next().await {
+        let done = match res {
+            Ok(rj) => rj?,
+            Err(err) => return Err(RunError::JobFailure(err)),
```
It seems this early return (and `should_quit` handling bellow) are fine w.r.t. cancel safety if `rewrite_commit` sets the `tokio::process::Command::kill_on_drop` which is not clear nor guaranteed it will be done. What I'm wondering is whether this invariant could be better enforced from `run_inner` directly, perhaps by constructing `Command` here (`run_inner` is already aware of using tokio so that shouldn't be a reservation)? The worry being that a later refactor might break the invariant.
